### PR TITLE
Implement ability to add row and/or column in the specified position

### LINKED ISSUE #9471: After filtering, insert data backwards
### Description

When I filter and insert data after a row, unfilter and find that the insert position is not after that row。

### Steps to reproduce

1.
2.
3.

### Demo

https://jsfiddle.net/handsoncode/dezfbgjt/

### Your environment
* Vue wrapper version:2.0
* Handsontable version:^11.1.0
* Browser name and version:chrome
* Operating system:windows

### PR DESCRIPTION
**Pull request by budnix**
### Context

The PR extends the Core `alter` method for the ability to add rows or columns in the specified position (_above_ or _below_) based on the given index.

##### Before
There is only one alter action `insert_row` that allows inserting rows only above the given index. For example `hot.alter('insert_row', 1)`.

##### After
The `insert_row` action stays for backward compatibility, but it's **deprecated** and works as previously. And the feature branch adds two new ones.
 *  `insert_row_above` it works the same as `insert_row` (and the action name is more explicit) but with one exception.
    *  Calling `insert_row` without arguments (`hot.alter('insert_row')`) currently adds row to the end of the table;
    * Calling `insert_row_above` without arguments (`hot.alter('insert_row_above')`) adds row to the beginning of the table;
 *  `insert_row_below` The new alter action allows adding new rows below the given index.

##### Before
There is only one alter action `insert_col` that allows inserting columns only on the left (or right in RTL) of the given index. For example `hot.alter('insert_col', 1)`.

##### After
The `insert_col` action stays for backward compatibility, but it's **deprecated** and works as previously. The PR adds two new ones.
 *  `insert_col_start` it works the same as `insert_col` (and the action name is more explicit) but with one exception.
    *  Calling `insert_col` without arguments (`hot.alter('insert_col')`) currently adds column to the most right edge of the table;
    * Calling `insert_col_start` without arguments (`hot.alter('insert_col_start')`) adds column to the most left edge of the table;
 *  `insert_col_end` The new alter action allows adding new columns on the right (or on the left in RTL).

Besides the above, the PR introduces the following changes:
 * Fix a bug in the `afterCreateRow` and `afterCreateCol` hooks. An incorrect visual column index revealed the error in the hook (the first argument). The visual index pointed to a record in the dataset before it was added to the IndexMapper (and dataset), and should point to the index after it was added;
 * Fix the bug where calling the above alter actions causes the selection misalignment (mostly when the whole table was selected);
 * Fix the `countEmptyCols` method that returns incorrect columns count when the rows are equal to `0`;

### How has this been tested?

I tested the changes locally and covered the feature and bug fixes with new E2E tests - including tests for TypeScript definitions.

### Types of changes

- [x] New feature or improvement (non-breaking change which adds functionality)

### Related issue(s):
1. fixes #9471

### Affected project(s):
- [x] `handsontable`

### Checklist:

- [x] I have reviewed the guidelines about [Contributing to Handsontable](https://github.com/handsontable/handsontable/blob/master/CONTRIBUTING.md) and I confirm that my code follows the code style of this project.
- [x] I have signed the [Contributor License Agreement](https://docs.google.com/forms/d/e/1FAIpQLScpMq4swMelvw3-onxC8Jl29m0fVp5hpf7d1yQVklqVjGjWGA/viewform?c=0&w=1)
- [x] My change requires a change to the documentation.

### CODE REVIEW

**Review by wszymanski — CHANGES_REQUESTED** [review]
Thanks for fixing bugs. It works as desired. I've found just one bug. The warning is called also for calling `insert_row_below`, `insert_row_above`, `insert_col_end` and `insert_col_start`.

File: handsontable/src/core.js [hunk 1]
```diff
@@ -530,6 +539,13 @@ export default function Core(rootElement, userSettings, rootInstanceSymbol = fal
         case 'insert_col': // backward compatibility
         case 'insert_col_start':
         case 'insert_col_end':
+
+          if (!deprecationWarns.has('insert_col')) {
+            deprecationWarns.add(action);
+            warn(toSingleLine`The alter action "insert_col" is deprecated and will be removed in the next\x20
+                              major release. Please use "insert_col_start" instead.`);
+          }
```
[the remark this answers is not in the document]
Comment (budnix) [reply handsontable/src/core.js]: Fixed within https://github.com/handsontable/handsontable/pull/9955/commits/1964a4f1547b4a827e8a9611952565771121791f

File: handsontable/src/core.js [hunk 2]
```diff
@@ -472,6 +475,12 @@ export default function Core(rootElement, userSettings, rootInstanceSymbol = fal
         case 'insert_row_below':
         case 'insert_row_above':
 
+          if (!deprecationWarns.has('insert_row')) {
+            deprecationWarns.add(action);
+            warn(toSingleLine`The alter action "insert_row" is deprecated and will be removed in the next\x20
+                              major release. Please use "insert_row_above" instead.`);
+          }
```
[the remark this answers is not in the document]
Comment (budnix) [reply handsontable/src/core.js]: Fixed within https://github.com/handsontable/handsontable/pull/9955/commits/1964a4f1547b4a827e8a9611952565771121791f

**Review by wszymanski — APPROVED** [review]

File: handsontable/src/core.js [hunk 3]
```diff
@@ -469,27 +471,48 @@ export default function Core(rootElement, userSettings, rootInstanceSymbol = fal
...
 
-            // Moving down the selection (when it exist). It should be present on the "old" row.
-            // TODO: The logic here should be handled by selection module.
-            if (isDefined(currentFromRow) && currentFromRow >= index) {
+            if (selection.isSelectedByCorner()) {
+              instance.selectAll();
+
+            } else if (isDefined(currentFromRow) && currentFromRow >= startVisualRowIndex) {
+              // Moving down the selection (when it exist). It should be present on the "old" row.
+              // TODO: The logic here should be handled by selection module.
```
Comment (kirszenbaum) [inline handsontable/src/core.js:515]:
```suggestion
              // Moving the selection (if it exists) downward – it should be applied to the "old" row.
              // TODO: The logic here should be handled by selection module.
```

File: handsontable/src/core.js [hunk 4]
```diff
@@ -501,36 +524,61 @@ export default function Core(rootElement, userSettings, rootInstanceSymbol = fal
...
+          }
+          // falls through
+        case 'insert_col_start':
+        case 'insert_col_end':
+          // "start" is a default behavior for creating new columns
+          const insertColumnMode = action === 'insert_col_end' ? 'end' : 'start';
+
+          // The line below adds support for backward compatibility "insert_col" alter action. Calling
+          // "insert_col" action without arguments adds new column on the right of the dataset. For "insert_col_start"
+          // action the column is added on the left of the dataset.
```
Comment (kirszenbaum) [inline handsontable/src/core.js:552]:
```suggestion
          // The line below ensures backward compatibility of the `alter()` method's `insert_col` action.
          // Calling the `insert_col` action with no arguments adds a new column to the right of the data set.
          // Calling the `insert_col_start` action adds a new column to the left of the data set.
```

File: handsontable/src/core.js [hunk 5]
```diff
@@ -501,36 +524,61 @@ export default function Core(rootElement, userSettings, rootInstanceSymbol = fal
...
+            const startVisualColumnIndex = instance.toVisualColumn(startColumnPhysicalIndex);
 
-            // Moving right the selection (when it exist). It should be present on the "old" row.
-            // TODO: The logic here should be handled by selection module.
-            if (isDefined(currentFromColumn) && currentFromColumn >= index) {
+            if (selection.isSelectedByCorner()) {
+              instance.selectAll();
+
+            } else if (isDefined(currentFromColumn) && currentFromColumn >= startVisualColumnIndex) {
+              // Moving right the selection (when it exist). It should be present on the "old" row.
```
Comment (kirszenbaum) [inline handsontable/src/core.js:580]:
```suggestion
              // Moving the selection (if it exists) rightward – it should be applied to the "old" column.
```

File: handsontable/src/core.js [hunk 6]
```diff
@@ -2588,8 +2643,12 @@ export default function Core(rootElement, userSettings, rootInstanceSymbol = fal
...
    * @param {string} action Possible alter operations:
    *  <ul>
-   *    <li> `'insert_row'` </li>
-   *    <li> `'insert_col'` </li>
+   *    <li> `'insert_row'` (alias for `'insert_row_above'` - deprecated) </li>
+   *    <li> `'insert_row_above'` </li>
+   *    <li> `'insert_row_below'` </li>
+   *    <li> `'insert_col'` (alias for `'insert_col_start'` - deprecated) </li>
+   *    <li> `'insert_col_start'` (inserts column at the left in LTR and at the right in RTL [layout direction](@/guides/internationalization/layout-direction.md) mode) </li>
+   *    <li> `'insert_col_end'` (inserts column at the right in LTR and at the left in RTL [layout direction](@/guides/internationalization/layout-direction.md) mode) </li>
```
Comment (kirszenbaum) [inline handsontable/src/core.js:2651]:
```suggestion
   *    <li> `'insert_row'` (<b>deprecated</b>, an alias for `'insert_row_above'`) </li>
   *    <li> `'insert_row_above'` </li>
   *    <li> `'insert_row_below'` </li>
   *    <li> `'insert_col'` (<b>deprecated</b>, an alias for `'insert_col_start'`) </li>
   *    <li> `'insert_col_start'` (depending on the [`layoutDirection`](@/api/options.md#layoutdirection), inserts a column to the left (in LTR) or to the right (in RTL) </li>
   *    <li> `'insert_col_end'` (depending on the [`layoutDirection`](@/api/options.md#layoutdirection), inserts a column to the right (in LTR) or to the left (in RTL) </li>
```

File: handsontable/src/core.js [hunk 7]
```diff
@@ -2601,9 +2660,11 @@ export default function Core(rootElement, userSettings, rootInstanceSymbol = fal
    * @example
    * ```js
    * // Insert new row above the row at given visual index.
-   * hot.alter('insert_row', 10);
+   * hot.alter('insert_row_above', 10);
+   * // Insert new row below the row at given visual index.
```
Comment (kirszenbaum) [inline handsontable/src/core.js:2664]:
```suggestion
   * // below the row with the given visual index, insert a new row
```

File: handsontable/src/dataMap/dataMap.js [hunk 8]
```diff
@@ -284,11 +284,13 @@ class DataMap {
    *
    * @param {number} [index] Physical index of the row before which the new row will be inserted.
    * @param {number} [amount=1] An amount of rows to add.
-   * @param {string} [source] Source of method call.
+   * @param {object} [options] Additional options for created rows.
+   * @param {string} [options.source] Source of method call.
+   * @param {'above'|'below'} [options.mode] Sets where the row will be inserted. Above or below the passed index.
```
Comment (kirszenbaum) [inline handsontable/src/dataMap/dataMap.js:289]:
```suggestion
   * @param {'above'|'below'} [options.mode] Sets where the row is inserted: above or below the passed index.
```

File: handsontable/src/dataMap/dataMap.js [hunk 9]
```diff
@@ -341,24 +343,36 @@ class DataMap {
...
 
   /**
    * Creates column at the right of the data array.
    *
    * @param {number} [index] Visual index of the column before which the new column will be inserted.
    * @param {number} [amount=1] An amount of columns to add.
-   * @param {string} [source] Source of method call.
+   * @param {object} [options] Additional options for created columns.
+   * @param {string} [options.source] Source of method call.
+   * @param {'start'|'end'} [options.mode] Sets where the column will be inserted. Start (left in LTR) or end (right in LTR)
```
Comment (kirszenbaum) [inline handsontable/src/dataMap/dataMap.js:370]:
```suggestion
   * @param {'start'|'end'} [options.mode] Sets where the column is inserted: at the start (left in [LTR](@/api/options.md#layoutdirection), right in [RTL](@/api/options.md#layoutdirection)) or at the end (right in LTR, left in LTR)
```

File: .changelogs/9471.json [hunk 10]
```diff
@@ -0,0 +1,7 @@
+{
+  "title": "Added a feature that allows adding row and column in the specified position. Above or Below the given index.",
```
Comment (kirszenbaum) [inline .changelogs/9471.json:2]:
```suggestion
  "title": "Added a new feature that lets you add rows and columns in a specified position (above or below the given index).",
```

**Review by kirszenbaum — CHANGES_REQUESTED** [review]

**Review by wszymanski — DISMISSED** [review]

**Review by kirszenbaum — APPROVED** [review]

**Comment by aninde** [comment]
@budnix I really like the huge amount of tests added and also include `layoutDirection: 'rtl'`, but I don't see any tests written for `manualColumnFreeze` or `fixedColumnsStart`. Should this feature not be supplemented by such?

**Pull request merged**